File: src/main/java/pl/yalgrin/playnite/simplesync/utils/MapperUtil.java
package pl.yalgrin.playnite.simplesync.utils;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import org.apache.commons.lang3.StringUtils;
import pl.yalgrin.playnite.simplesync.dto.AbstractObjectDTO;
import pl.yalgrin.playnite.simplesync.dto.LinkDTO;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class MapperUtil {

    public static boolean hasChanged(Long obj1, Long obj2) {
        return !Objects.equals(Optional.ofNullable(obj1).orElse(0L), Optional.ofNullable(obj2).orElse(0L));
    }

    public static boolean hasChanged(String obj1, String obj2) {
        return !StringUtils.equals(obj1, obj2);
    }

    public static boolean hasChanged(Object obj1, Object obj2) {
        return !Objects.equals(obj1, obj2);
    }

    public static boolean hasChanged(List<? extends AbstractObjectDTO> oldList,
                                     List<? extends AbstractObjectDTO> newList) {
        Set<String> oldIds = Optional.ofNullable(oldList)
                .map(e -> e.stream().map(AbstractObjectDTO::getId).filter(Objects::nonNull).collect(Collectors.toSet()))
                .orElseGet(Collections::emptySet);
        Set<String> newIds = Optional.ofNullable(newList)
                .map(e -> e.stream().map(AbstractObjectDTO::getId).filter(Objects::nonNull).collect(Collectors.toSet()))
                .orElseGet(Collections::emptySet);
        return oldIds.size() != newIds.size() || oldIds.stream().anyMatch(id -> !newIds.contains(id));
    }

    public static boolean haveLinksChanged(List<LinkDTO> oldList, List<LinkDTO> newList) {
        oldList = Optional.ofNullable(oldList).orElseGet(Collections::emptyList);
        newList = Optional.ofNullable(newList).orElseGet(Collections::emptyList);
        return oldList.size() != newList.size() || anyLinkChanged(oldList, newList);
    }

    private static boolean anyLinkChanged(List<LinkDTO> oldList, List<LinkDTO> newList) {
        return IntStream.range(0, oldList.size()).anyMatch(i -> {
            LinkDTO oldLink = oldList.get(i);
            LinkDTO newLink = newList.get(i);
            return !StringUtils.equals(oldLink.getUrl(), newLink.getUrl()) || !StringUtils.equals(oldLink.getName(),
                    newLink.getName());
        });
    }

}

File: src/main/kotlin/pl/yalgrin/playnite/simplesync/common/util/MapperUtil.kt
package pl.yalgrin.playnite.simplesync.common.util

import org.apache.commons.lang3.Strings
import pl.yalgrin.playnite.simplesync.library.dto.LibraryObjectDTO
import pl.yalgrin.playnite.simplesync.library.dto.LinkDTO

object MapperUtil {
    fun hasChanged(obj1: Long?, obj2: Long?): Boolean {
        return (obj1 ?: 0L) != (obj2 ?: 0L)
    }

    fun hasChanged(obj1: String?, obj2: String?): Boolean {
        return !Strings.CS.equals(obj1, obj2)
    }

    fun hasChanged(obj1: Any?, obj2: Any?): Boolean {
        return obj1 != obj2
    }

    fun hasChanged(
        oldList: List<LibraryObjectDTO>?,
        newList: List<LibraryObjectDTO>?
    ): Boolean {
        val oldIds = oldList?.mapNotNull { it.id }?.toSet().orEmpty()
        val newIds = newList?.mapNotNull { it.id }?.toSet().orEmpty()
        return oldIds.size != newIds.size || oldIds.any { it !in newIds }
    }

    fun haveLinksChanged(oldList: List<LinkDTO>?, newList: List<LinkDTO>?): Boolean {
        val oldLinks = oldList.orEmpty()
        val newLinks = newList.orEmpty()
        return oldLinks.size != newLinks.size || oldLinks.indices.any { i ->
            val oldLink = oldLinks[i]
            val newLink = newLinks[i]
            !Strings.CS.equals(oldLink.url, newLink.url) || !Strings.CS.equals(oldLink.name, newLink.name)
        }
    }
}